src/Surfnet/StepupSelfService/SelfServiceBundle/Service/SmsSecondFactor/SessionChallengeStore.php

```diff
@@ -52,18 +52,16 @@ public function generateChallenge()
         };
         $challenge = join('', array_map('chr', array_map($randomCharacters, range(1, 8))));
 
-        $this->session->set($this->sessionKey, $challenge);
+        $this->session->set($this->sessionKey . '/' . $challenge, $phoneNumber);
 
         return $challenge;
     }
 
-    public function verifyChallenge($challenge)
+    public function takePhoneNumberMatchingChallenge($challenge)
     {
         $challenge = strtoupper($challenge);
-        $expectedChallenge = $this->session->get($this->sessionKey);
+        $phoneNumber = $this->session->remove($this->sessionKey . '/' . $challenge);
```

Comment:
Hrms, seems the SessionBagInterface doesn't contain remove :panda_face: 

Nitpick: Why `takeXXX` by the way? Doesn't seem to match with the `generateChallenge()` naming style...